### PR TITLE
An experiment to substitute onMeasureChanged with ResizeObserver

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #31540

TODO:

- [ ] Tests

### CODE REVIEW

File: src/service/resize-observer.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  /**
+   * @param {!Array<!ResizeObserverEntry>} entries
+   * @private
+   */
+  processEntries_(entries) {
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, clientRect} = entries[i];
+      let seenTarget = false;
```
Comment (samouri) [inline src/service/resize-observer.js:117]: this is to ensure it only gives the latest size? are you confident about the ordering of the array? (newest first)

File: src/service/resize-observer.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      const callbacks = this.callbacks_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      const {width, height} = clientRect;
+      /** @type {../layout-rect.LayoutSizeDef} */
+      const size = {width, height};
+      this.lastSizes_.set(target, size);
+      for (let k = 0; k < callbacks.length; k++) {
+        callCallbackNoInline(callbacks[k], size);
```
Comment (samouri) [inline src/service/resize-observer.js:136]: nit: this is cleaner imo if you pass in the whole thunk. E.g. 
```js
callNoInline(() => callbacks[k](size))
```

File: src/service/resize-observer.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    let callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      callbacks = [];
+      this.callbacks_.set(element, callbacks);
+      this.observer_.observe(element);
+    }
+    if (!pushIfNotExist(callbacks, callback)) {
+      return;
+    }
+    const lastSize = this.lastSizes_.get(element);
```
Comment (samouri) [inline src/service/resize-observer.js:84]: Can we ensure the `lastSize` logic only occurs in the case of an element with a connected concurrent observer.
I want to protect from a very stale value being used and then quickly corrected after observation. 

E.g.:
1. Observe el1. After some time, unobserve el1
2. While not observed, size changes happen to el1
3. Observe el1 again. The stale `lastSize` is used initially.

File: src/service/resize-observer.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ObserverCallbackDef;
+
+/**
+ * Pools resize observers. Eventually the hope is that the call sites can
+ * instantiate their own resize observers. But currently there are some
+ * performance issues and pooling is desired. For more info, see
+ * https://bugs.chromium.org/p/chromium/issues/detail?id=1157544
```
Comment (samouri) [inline src/service/resize-observer.js:31]: Thanks for linking this bug! Very interesting results with a significant impact on how to structure our ResOb code.

**Comment by samouri** [comment]
Shouldn't we merge the ResizeObserver polyfill before making the changes in `extensions`? Its fine to create the new helper service behind a server-side flag though

File: src/service/resize-observer.js (hunk 1 above)
Comment (dvoytenko) [reply src/service/resize-observer.js:117]: The ordering is per spect. But the newest are last. That's why I'm going through array in the reverse order.

File: src/service/resize-observer.js (hunk 2 above)
Comment (dvoytenko) [reply src/service/resize-observer.js:136]: I was trying to reduce number of function allocations.

File: src/service/resize-observer.js (hunk 3 above)
Comment (dvoytenko) [reply src/service/resize-observer.js:84]: I clean out the `lastSize` after the `unobserve` call, thus there shouldn't be stale values here.

**Comment by dvoytenko** [comment]
> Shouldn't we merge the ResizeObserver polyfill before making the changes in extensions? Its fine to create the new helper service behind a server-side flag though

I added the build flag just for this reason. Polyfill itself is pretty easy and we can do it as early as we want. But the polyfill suffers from the similar issues that we've previously fixed for `IntersectionObserver` - namely working across the boundary of a shadow root and a same-origin iframe boundary.

**Comment by dvoytenko** [comment]
Closed in favor of https://github.com/ampproject/amphtml/pull/31899

**Pull request closed without merging**